Runtime project refactorization

pluskal opened the pull request:
As a preparation for [Adding async support](https://github.com/kaitai-io/kaitai_struct/issues/640) I propose following refactorizations:
* Add EditorConfig with default MS configuration.
  * Extracted from [docs](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019), feel free to propose any changes
* Split types to separate files
* Extract IKaitaiStream interface.
* Use composition instead of inheritance with BinaryReader

pluskal committed:
Add EditorConfig with default MS configuration.

pluskal committed:
Split types to separate files

pluskal committed:
Extract IKaitaiStream interface.

pluskal committed:
Use composition instead of inheritance with BinaryReader

KOLANICH commented on Exception/ValidationFailedError.cs:
```diff
@@ -0,0 +1,36 @@
+        protected KaitaiStream io;
+
+        public ValidationFailedError(string msg, KaitaiStream io, string srcPath)
+            : base("at pos " + io.Pos + ": validation failed: " + msg, srcPath)
```
IMHO: positions in the stream must be stored as a number field.

pluskal replied:
Let's improve this in follow-ups.

KOLANICH commented on Exception/ValidationNotEqualError.cs:
```diff
@@ -0,0 +1,30 @@
+        protected Object expected;
+
+        public ValidationNotEqualError(byte[] expected, byte[] actual, KaitaiStream io, string srcPath)
+            : base("not equal, expected " + ByteArrayToHex(expected) + ", but got " + ByteArrayToHex(actual), io,
```
The same problem. All the stuff must be stored as a field.

KOLANICH commented on Interface/CustomDecoder.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    ///     A custom decoder interface. Implementing classes can be called from
+    ///     inside a .ksy file using `process: XXX` syntax.
+    /// </summary>
```
Take a look at my new architecture for python runtime. https://gitlab.com/KOLANICH/Cache.py/tree/master/Cache/kaitai/compress . It may worth to mirror it here.

pluskal replied:
Let's improve this in follow-ups.

pluskal commented:
@KOLANICH thanks for comments, I will look into it. But, this PR is meant only as cleanup/refactorization.

I do not want to change functionality in this PR, only to change a code structure to make it cleaner for future PRs with new functionality/improvements. 

At least, that is what I thought that @GreyCat meant in this [comment](https://github.com/kaitai-io/kaitai_struct/issues/640#issuecomment-550896430) 
> "Let's go in the small steps and do these as separate PRs"

GreyCat commented on .editorconfig:
```diff
@@ -0,0 +1,153 @@
+
+[*.vb]
+# Modifier preferences
+visual_basic_preferred_modifier_order = Partial,Default,Private,Protected,Public,Friend,NotOverridable,Overridable,MustOverride,Overloads,Overrides,MustInherit,NotInheritable,Static,Shared,Shadows,ReadOnly,WriteOnly,Dim,Const,WithEvents,Widening,Narrowing,Custom,Async:suggestion
```
Missing EOF newline here. I know that some Microsoft tool insist on having no trailing EOF newlines, but editorconfig should be universal thing, not relying on Microsoft practices.

pluskal replied:
EOL Added

GreyCat commented on .editorconfig:
```diff
@@ -0,0 +1,153 @@
+﻿###############################
+# Core EditorConfig Options   #
+###############################
```
I believe this file was mirrored from somewhere? If it is, let us add an URL to original copy in the comments, so further generations of developers could continue to update it from the source?

pluskal replied:
Source added

GreyCat commented on Exception/KaitaiStructError.cs:
```diff
@@ -0,0 +1,20 @@
+    ///     Common ancestor for all error originating from Kaitai Struct usage.
+    ///     Stores KSY source path, pointing to an element supposedly guilty of
+    ///     an error.
+    /// </summary>
```
I've noticed that you indent XML comments with one extra 4-space indent in `<summary>...</summary>` — can you point me to some kind of official standard regarding that?

pluskal replied:
Nice catch! I use Resharper and it does that by default. https://stackoverflow.com/questions/16533233/resharper-indenting-the-xml-documentation-text-on-code-cleanup

Reverted.

GreyCat commented:
Thanks for taking time to separate the original proposal and maintain cleaner commit history!
I've added some (minor) comments / asks, please take a look?

pluskal committed:
Add source and EOL in .editorconfig

pluskal committed:
Fix XML comment indentation.

pluskal committed:
Change file layout -> fileds, constructors, properties. Chage abstract class constructor visibility to protected.

pluskal committed:
Use string interpolation and clean up redundant code.

pluskal commented:
Sorry for taking so long to fix it. I addressed your comments.

With the commit `Use string interpolation and clean up redundant code.` I run the code through the Resharper code cleanup to remove redundant code and use string interpolation. 

Let me know what do you think of that, it is just a proposal that can be removed.

pluskal commented:
@GreyCat can you check this PR out please. I would like to continue with `async` support.
